main.go

```diff
@@ -0,0 +1,341 @@
+		setupLog.Error(err, "unable to create controller", "controller", "UpgradeConfig")
+		os.Exit(1)
+	}
+
+	// Add NodeKeeper controller to the manager
+	if err = (&nodekeeper.ReconcileNodeKeeper{
+		Client:                      mgr.GetClient(),
+		Scheme:                      mgr.GetScheme(),
+		ConfigManagerBuilder:        configmanager.NewBuilder(),
+		Machinery:                   machinery.NewMachinery(),
+		MetricsClientBuilder:        metrics.NewBuilder(),
+		DrainstrategyBuilder:        drain.NewBuilder(),
+		UpgradeConfigManagerBuilder: upgradeconfigmanager.NewBuilder(),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "NodeKeeper")
+		os.Exit(1)
+	}
+
+	// Add MachineConfigPool controller to the manager
+	if err = (&machineconfigpool.ReconcileMachineConfigPool{
+		Client:                      mgr.GetClient(),
+		Scheme:                      mgr.GetScheme(),
+		UpgradeConfigManagerBuilder: ucm.NewBuilder(),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "MachineConfigPool")
+		os.Exit(1)
+	}
+
+	ctx := context.TODO()
+
+	// Get a config to talk to the apiserver
+	cfg, err := config.GetConfig()
+	if err != nil {
+		log.Error(err, "")
+		os.Exit(1)
+	}
+
+	// Add the Metrics Service and ServiceMonitor
+	if err := addMetrics(ctx, cfg); err != nil {
+		log.Error(err, "Metrics service is not added.")
+		os.Exit(1)
+	}
+
+	// Add the Custom Metrics Service
+	metricsClient, err := client.New(cfg, client.Options{})
+	if err != nil {
+		log.Error(err, "unable to create k8s client for upgrade metrics")
+		os.Exit(1)
+	}
+
+	uCollector, err := collector.NewUpgradeCollector(metricsClient)
+	if err != nil {
+		log.Error(err, "unable to create upgrade metrics collector")
+		os.Exit(1)
+	}
+
+	ns, err := util.GetOperatorNamespace()
+	if err != nil {
+		os.Exit(1)
+	}
+
+	customMetrics := opmetrics.NewBuilder(ns, "managed-upgrade-operator-custom-metrics").
+		WithPath(customMetricsPath).
+		WithCollector(uCollector).
+		WithServiceMonitor().
+		WithServiceLabel(map[string]string{"name": muocfg.OperatorName}).
+		GetConfig()
+
+	if err = opmetrics.ConfigureMetrics(context.TODO(), *customMetrics); err != nil {
+		log.Error(err, "Failed to configure custom metrics")
+		os.Exit(1)
+	}
+
+	//+kubebuilder:scaffold:builder
+
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up health check")
+		os.Exit(1)
+	}
+	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up ready check")
+		os.Exit(1)
+	}
+
+	// Define stopCh which we'll use to notify the upgradeConfigManager (and any other routine)
+	// to stop work. This channel can also be used to signal routines to complete any cleanup
+	// work
+	stopCh := signals.SetupSignalHandler()
+
+	upgradeConfigManagerClient, err := client.New(cfg, client.Options{})
```

Comment:
As this is setting up its own client it does not have the upgradev1alpha1 types added its scheme. Currently failing to test as a result of that.

I thought it could be supplied with the `mgr.GetClient()` client instead rather than using its own, so I tried that and tested it, but it didn't work. Have not looked deeper into what's happening yet.

Reply:
Okay, it actually does work to use `mgr.GetClient()` I think, it's just that there's a similar issue elsewhere that was triggering the same problem and I got caught up on assuming that it was still this bit of code. I will comment on the other place it needs addressing.

Reply:
I ended up going with the separate client and doing the following instead - seems to work ok:
```suggestion
	upgradeConfigManagerClient, err := client.New(cfg, client.Options{Scheme: mgr.GetScheme()})
```

Reply:
Done :+1: